xarray/util/generate_reductions.py

```diff
@@ -5,292 +5,389 @@
 Usage:
     python xarray/util/generate_reductions.py > xarray/core/_reductions.py
     pytest --doctest-modules xarray/core/_reductions.py --accept || true
-    pytest --doctest-modules xarray/core/_reductions.py --accept
+    pytest --doctest-modules xarray/core/_reductions.py
 
 This requires [pytest-accept](https://github.com/max-sixty/pytest-accept).
 The second run of pytest is deliberate, since the first will return an error
 while replacing the doctests.
 
 """
-
 import collections
 import textwrap
-from functools import partial
-from typing import Callable, Optional
+from dataclasses import dataclass
 
 MODULE_PREAMBLE = '''\
 """Mixin classes with reduction operations."""
 # This file was generated using xarray.util.generate_reductions. Do not edit manually.
 
-import sys
-from typing import Any, Callable, Hashable, Optional, Sequence, Union
+from typing import TYPE_CHECKING, Any, Callable, Hashable, Optional, Sequence, Union
 
 from . import duck_array_ops
-from .types import T_DataArray, T_Dataset
 
-if sys.version_info >= (3, 8):
-    from typing import Protocol
-else:
-    from typing_extensions import Protocol'''
+if TYPE_CHECKING:
+    from .dataarray import DataArray
+    from .dataset import Dataset'''
+
+
+CLASS_PREAMBLE = """
 
-OBJ_PREAMBLE = """
+class {obj}{cls}Reductions:
+    __slots__ = ()
 
-class {obj}Reduce(Protocol):
     def reduce(
         self,
         func: Callable[..., Any],
         dim: Union[None, Hashable, Sequence[Hashable]] = None,
+        *,
         axis: Union[None, int, Sequence[int]] = None,
         keep_attrs: bool = None,
         keepdims: bool = False,
         **kwargs: Any,
-    ) -> T_{obj}:
-        ..."""
+    ) -> "{obj}":
+        return NotImplemented"""
```

Comment:
```suggestion
        raise NotImplementedError()"""
```
Raising an error for this dummy method looks a bit better.